### DISCUSSION TITLE
msvc with clangd

### DISCUSSION THREAD

**Discussion by RI-ss** [clangd/clangd#1573] [Q&A]
i use xmake to build a project in vscode
xmake can generate compile_commands for clangd
but here is a problem when toolchain is msvc

when the toolchain is clang,clangd work correctly
![image](https://user-images.githubusercontent.com/71321125/229559759-767f5357-bcb8-4a36-9110-c4089fcce807.png)
![image](https://user-images.githubusercontent.com/71321125/229560625-7bdf9a07-a35a-4806-8632-53440d26da9c.png)
![image](https://user-images.githubusercontent.com/71321125/229560676-d23178df-5ce0-4e46-a95d-308b1fb752a3.png)
![image](https://user-images.githubusercontent.com/71321125/229560717-8e669fa8-5ccc-4441-85be-1c538d4315ab.png)

but i dont want to use clang, 
when toolchain is msvc ,there are some problems
compile_commands changes to this
![image](https://user-images.githubusercontent.com/71321125/229561716-800e4630-415a-4133-84fa-4f4c330906b5.png)
clangd starts to show errors
![image](https://user-images.githubusercontent.com/71321125/229562589-ce6dfb5c-4532-47bb-bef5-a14b928177a1.png)
and exchange the line 1 and line 2
iostream cannt be found
![image](https://user-images.githubusercontent.com/71321125/229562781-715dd6ee-6564-42a8-be45-e32a3c225fa8.png)

the project are able to be built and run when toolchain is msvc , can not build correctly  in clang

the main difference is the first element of argument,chang->cl,and it is automatically generated by xmake
is this generating right?
if it is right,what shold i to do to make it work?
if it is worng,what kind is right?

**Comment by i-ky** [comment]
I would suggest adding MSVC executable to `--query-driver` in `clangd.arguments` in VS Code settings. This will grant clangd permission to invoke compiler for the purpose of standard include path extraction.

**Reply by RI-ss** [reply]
![image](https://user-images.githubusercontent.com/71321125/229701089-8f4632c0-989a-497b-a097-81e6b1d6e325.png)
like this?
but it doesnt work

**Reply by i-ky** [reply]
In your `compile_commands.json` it says `...\cl.exe`, so `query-driver` should be `...\cl.exe` too. There is a simple text based match algorithm behind.

Wih `--log=verbose` you should be able to troubleshoot, when clangd decides to invoke compiler and when not.

**Reply by HighCommander4** [reply]
I don't know if `cl.exe` supports the same flags (e.g. `<driver> -E -v -x c++ -`) that clangd uses to try to extract built-in includes from the driver.

**Reply by RI-ss** [reply]
![image](https://user-images.githubusercontent.com/71321125/229754067-981674e8-c7bf-4427-b0ee-3ca0d3029458.png)
![image](https://user-images.githubusercontent.com/71321125/229754126-1efb0460-a79f-4628-8bce-b280e0c24b9c.png)
ok,nothing changed

there is the output of clangd

I[18:42:19.443] <-- shutdown(8)
I[18:42:19.443] --> reply:shutdown(8) 0 ms
I[18:42:19.490] <-- exit
I[18:42:19.490] LSP finished, exiting with status 0
The flag `-clang-tidy-checks` is obsolete and ignored.
I[18:42:19.530] clangd version 15.0.7
I[18:42:19.532] Features: windows
I[18:42:19.532] PID: 10776
I[18:42:19.532] Working directory: d:/c++/test2
I[18:42:19.532] argv[0]: C:\msys64\mingw64\bin\clangd.exe
I[18:42:19.532] argv[1]: --background-index
I[18:42:19.532] argv[2]: --compile-commands-dir=.vscode
I[18:42:19.532] argv[3]: -j=13
I[18:42:19.532] argv[4]: --clang-tidy
I[18:42:19.532] argv[5]: --clang-tidy-checks=performance-*,bugprone-*
I[18:42:19.532] argv[6]: --all-scopes-completion
I[18:42:19.532] argv[7]: --completion-style=detailed
I[18:42:19.532] argv[8]: --header-insertion=iwyu
I[18:42:19.532] argv[9]: --pch-storage=memory
I[18:42:19.532] argv[10]: --header-insertion=never
I[18:42:19.532] argv[11]: --query-driver=C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\bin\HostX64\x64\cl.exe
I[18:42:19.536] Starting LSP over stdin/stdout
I[18:42:19.537] <-- initialize(0)
I[18:42:19.539] --> reply:initialize(0) 2 ms
I[18:42:19.542] <-- initialized
I[18:42:19.551] <-- textDocument/didOpen
I[18:42:19.553] Loaded compilation database from d:/c++/test2/.vscode/compile_commands.json
I[18:42:19.553] --> window/workDoneProgress/create(0)
I[18:42:19.554] Enqueueing 1 commands for indexing
I[18:42:19.555] <-- reply(0)
I[18:42:19.555] --> $/progress
I[18:42:19.555] --> $/progress
I[18:42:19.563] <-- textDocument/documentSymbol(1)
I[18:42:19.564] <-- textDocument/codeAction(2)
I[18:42:19.611] <-- textDocument/documentLink(3)
I[18:42:19.612] <-- textDocument/inlayHint(4)
I[18:42:19.752] --> $/progress
I[18:42:19.753] --> $/progress
E[18:42:19.836] System include extraction: driver execution failed with return code: 2 - ''. Args: ["C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.34.31933\\bin\\HostX64\\x64\\cl.exe" -E -x c++ - -v]
I[18:42:19.837] ASTWorker building file d:/c++/test2/src/main.cpp version 6 with command 
[d:\c++\test2]
"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe" --driver-mode=cl /c /nologo /O2 /fp:fast /std:c++20 "/Idependencies\\include" /EHsc "/ID:\\Program Files\\vcpkg\\installed\\x64-windows-static\\include" /DNDEBUG "/Fobuild\\.objs\\test2\\windows\\x64\\release\\src\\main.cpp.obj" -resource-dir=C:/msys64/mingw64/lib/clang/15.0.7 -- d:/c++/test2/src/main.cpp
I[18:42:19.844] --> textDocument/clangd.fileStatus
I[18:42:20.317] --> workspace/semanticTokens/refresh(1)
I[18:42:20.318] <-- reply(1)
I[18:42:20.338] --> textDocument/publishDiagnostics
I[18:42:20.338] --> reply:textDocument/documentSymbol(1) 775 ms
I[18:42:20.339] --> reply:textDocument/codeAction(2) 774 ms
I[18:42:20.339] --> reply:textDocument/documentLink(3) 727 ms
I[18:42:20.339] --> reply:textDocument/inlayHint(4) 726 ms
I[18:42:20.339] --> textDocument/clangd.fileStatus
I[18:42:20.604] <-- textDocument/codeAction(5)
I[18:42:20.604] --> reply:textDocument/codeAction(5) 0 ms
I[18:42:20.604] --> textDocument/clangd.fileStatus
I[18:42:22.902] <-- textDocument/inlayHint(6)
I[18:42:22.902] --> reply:textDocument/inlayHint(6) 0 ms
I[18:42:22.902] --> textDocument/clangd.fileStatus
I[18:42:29.689] <-- textDocument/codeAction(7)
I[18:42:29.691] --> reply:textDocument/codeAction(7) 1 ms
I[18:42:29.691] --> textDocument/clangd.fileStatus

**Reply by i-ky** [reply]
As @HighCommander4 said, MSVC does not support flags that clangd is using for include path extraction:
```
E[18:42:19.836] System include extraction: driver execution failed with return code: 2 - ''. Args: ["C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\bin\HostX64\x64\cl.exe" -E -x c++ - -v]
```

**Comment by HighCommander4** [comment]
Is there something preventing you from using a `compile_commands.json` generated with a clang toolchain, even as you use msvc for building?

**Reply by RI-ss** [reply]
no, 
it is ok that all are accomplished in clang,
but i just am curious to find a way to use clangd in msvc,rather than code in clang and build in msvc

**Comment by bowenxuuu** [comment]
clangd could not query the standard library header path from msvc ( maybe a bug ? ).

So, you can add one line into the CMakeLists.txt of your project, like this:

include_directories("C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\include")

**Reply by bowenxuuu** [reply]
<img width="924" alt="11" src="https://github.com/clangd/clangd/assets/40262910/70e47f5e-0e3d-4567-8442-46dce87c1ecb">
<img width="1061" alt="22" src="https://github.com/clangd/clangd/assets/40262910/92f699eb-5561-43f5-b34e-6942fa97b7db">
<img width="1279" alt="33" src="https://github.com/clangd/clangd/assets/40262910/c670aeaa-d005-4868-b8d2-df8a23778548">

**Reply by RI-ss** [reply]
![image](https://github.com/clangd/clangd/assets/71321125/db91c841-52dc-4984-9bad-f25b0dc801f7)
![image](https://github.com/clangd/clangd/assets/71321125/40d43dab-7fec-4fb1-8dfa-db8798d40a54)
![image](https://github.com/clangd/clangd/assets/71321125/f7f04fdc-9dd3-4759-a667-1cdb09c39b60)
![image](https://github.com/clangd/clangd/assets/71321125/bfb12db6-6d81-4ab3-a6e5-251301b0fa97)
![image](https://github.com/clangd/clangd/assets/71321125/5faaff89-949a-4d02-b02c-11e945b7e74c)

确实是可以手动添加vs头
但是并不能正确识别vc的库

**Reply by laugh12321** [reply]
@RI-ss 若想要在VScode中使用Clangd用于MSVC项目提示，仅需在安装Visual Studio时，可选组件下选择“适用于 Windows 的 C++ Clang 编译器”来安装即可。

在VScode配置中添加clangd路径
```
"clangd.path": "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/Llvm/bin/clangd.exe",
```
或者将`C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\Llvm\bin`添加至PATH即可

**Reply by bowenxuuu** [reply]
装 vs2022吧，我之前就是用的 vs2019，clangd 死活不能自动识别标准库。换成 vs2022 之后，无需任何手动添加的额外配置，clangd 自动完美识别。vs2022里面一定要装 C++ Build Insights 这个组件，它的作用是提高 MSVC 工具链的可见性。
![微信截图_20240515091200](https://github.com/clangd/clangd/assets/40262910/d5eb42fa-7876-4ad1-9cc3-ed17b4724bf3)
![v2-392fafc1ec2e5f4ed8d72eba4ae6ebdb_1440w](https://github.com/clangd/clangd/assets/40262910/17d298b8-42ef-4756-a5e0-0503f2f9b774)

**Reply by rocascent** [reply]
> 装 vs2022吧，我之前就是用的 vs2019，clangd 死活不能自动识别标准库。换成 vs2022 之后，无需任何手动添加的额外配置，clangd 自动完美识别。vs2022里面一定要装 C++ Build Insights 这个组件，它的作用是提高 MSVC 工具链的可见性。 

牛的。不过如果设置为`/std:c++latest`会找不到某些标准库，`/std:c++20`就行

**Reply by zse4t7i** [reply]
> 牛的。不过如果设置为`/std:c++latest`会找不到某些标准库，`/std:c++20`就行

```.clangd
//.clangd
CompileFlags:
    Add: ["/clang:-std=c++23"]
```
这可能有用，似乎是因为 clangd 使用 clang 来分析错误，而 clang 没有c++latest这个参数，只有c++23。具体参见https://github.com/clangd/clangd/issues/1850#issuecomment-2332356041。

**Comment by hxhc** [comment]
> 确实是可以手动添加vs头 但是并不能正确识别vc的库

可以把MSVC的include目录和lib目录放在环境变量里，虽然这样不太优雅

**Comment by Touchmyf1sh** [comment]
现在我和你们的问题相反，无论用什么编译器，头文件跳转只能到MSVC的头文件，就算定义了--query-driver到gcc的编译器还是一样的,互联网上找了一圈没找到好的解决办法。装的是VS2022,也装了c++ build insights,难道是这东西把MSVC的可见性提高到很高的位置吗
![image](https://github.com/user-attachments/assets/e5ef9d8f-e5a3-4b6c-94a3-de81835f5679)
![image](https://github.com/user-attachments/assets/6e65871c-6ba0-4471-ba93-cde7ec2fb727)

**Reply by owlx56** [reply]
@Touchmyf1sh 卸载 c++ build insights 会怎么样？

或者修改 `config.yaml`/`.clangd` 配置文件：https://github.com/clangd/clangd/issues/1938#issuecomment-1953686357
